transports/quic/tests/smoke.rs

```diff
@@ -68,614 +31,408 @@ async fn async_std_smoke() {
-            Some(SwarmEvent::Behaviour(RequestResponseEvent::Message {
-                message:
-                    RequestResponseMessage::Request {
-                        request: Ping(data),
-                        channel,
-                        ..
-                    },
-                ..
-            })) => {
-                b.behaviour_mut()
-                    .send_response(channel, Pong(data))
-                    .unwrap();
-            }
-            e => panic!("{:?}", e),
-        }
+    let quic_transport = quic::async_std::Transport::new(config);
+    let tcp_transport = tcp::async_io::Transport::new(tcp::Config::default())
+        .upgrade(upgrade::Version::V1)
+        .authenticate(
+            noise::NoiseConfig::xx(
+                noise::Keypair::<noise::X25519Spec>::new()
+                    .into_authentic(&keypair)
+                    .unwrap(),
+            )
+            .into_authenticated(),
+        )
+        .multiplex(yamux::YamuxConfig::default());
 
-        match b.next().await {
-            Some(SwarmEvent::Behaviour(RequestResponseEvent::ResponseSent { .. })) => {}
-            e => panic!("{:?}", e),
-        }
+    let transport = OrTransport::new(quic_transport, tcp_transport)
+        .map(|either_output, _| match either_output {
+            EitherOutput::First((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+            EitherOutput::Second((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+        })
+        .boxed();
 
-        sync_tx.send(()).unwrap();
+    (peer_id, transport)
+}
 
-        match b.next().await {
-            Some(SwarmEvent::ConnectionClosed {
-                cause: Some(ConnectionError::IO(_)),
-                ..
-            }) => {}
-            e => panic!("{:?}", e),
-        }
-    };
+#[cfg(feature = "async-std")]
+#[async_std::test]
+async fn tcp_and_quic() {
+    let (a_peer_id, mut a_transport) = new_tcp_quic_transport();
+    let (b_peer_id, mut b_transport) = new_tcp_quic_transport();
 
-    join(fut_a, fut_b).await;
-}
+    let quic_addr = start_listening(&mut a_transport, "/ip4/127.0.0.1/udp/0/quic").await;
+    let tcp_addr = start_listening(&mut a_transport, "/ip4/127.0.0.1/tcp/0").await;
 
-#[derive(Debug, Clone)]
-struct PingProtocol();
+    let ((a_connected, _, _), (b_connected, _)) =
+        connect(&mut a_transport, &mut b_transport, quic_addr).await;
+    assert_eq!(a_connected, b_peer_id);
+    assert_eq!(b_connected, a_peer_id);
 
-#[derive(Clone)]
-struct PingCodec();
+    let ((a_connected, _, _), (b_connected, _)) =
+        connect(&mut a_transport, &mut b_transport, tcp_addr).await;
+    assert_eq!(a_connected, b_peer_id);
+    assert_eq!(b_connected, a_peer_id);
+}
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-struct Ping(Vec<u8>);
+// Note: This test should likely be ported to the muxer compliance test suite.
+#[cfg(feature = "async-std")]
+#[test]
+fn concurrent_connections_and_streams_async_std() {
```

Comment:
This test hangs forever when I run it locally.

Reply:
Are you sure it hangs forever and doesn't just take long? I had this test take up to 90s some times locally although that didn't happen recently.

Reply:
Found the issue by enabling the `nocapture` flag: The initial handshake timed out, which caused the spawned tasks to panic in l. 279 and l. 260. Because this happens in a separate task the test itself did not panic. Setting the handshake timeout to 5s fixes it.

It would be great if the test could panic if one of the spawned tasks panics. Maybe simply by having something like a oneshot channel for each task so we get notified if it dropped?

Reply:
If we don't spawn separate tasks, then the test is not actually concurrent!

I added a timeout to the final `block_on` which should trigger if this happens again.